### ISSUE TITLE
[FEA] cudf directory reorganization

### ISSUE THREAD

**Issue by shwina** [rapidsai/cudf#1983]
cuDF can use a little bit of directory reorganization: specifically, many modules in `cudf/dataframe` could find better homes. For example:

* `buffer.py`, `column.py`, `columnops.py` etc., could be relocated to a `cudf.internals` (or better still `cudf._internals`) package. I'd propose something like:

```
cudf
└── _internals
    ├── __init__.py
    ├── buffer.py
    └── column
        ├── __init__.py
        ├── categorical.py
        ├── column.py
        ├── columnops.py
        ├── datetime.py
        ├── numerical.py
        └── string.py
```

* `multiindex.py` and `cudf/indexing.py` could be relocated to a new `cudf.indexing` package

* `series.py`, `dataframe.py` and `index.py` could be relocated to a new `cudf.core` package

I'd be happy to provide an initial directory structure for us to iterate on. Any suggestions welcome!

**Comment by shwina** [comment]
Here is a proposed directory structure that somewhat mimics Pandas:

```
cudf/
├── __init__.py
├── bindings/
├── core/
│   ├── __init__.py
│   ├── _internals/
│   │   ├── __init__.py
│   │   ├── buffer.py
│   │   └── column/
│   │       ├── __init__.py
│   │       ├── categorical.py
│   │       ├── column.py
│   │       ├── columnops.py
│   │       ├── datetime.py
│   │       ├── numerical.py
│   │       └── string.py
│   ├── dataframe.py
│   ├── groupby/
│   ├── index/
│   │   ├── __init__.py
│   │   ├── index.py
│   │   └── multi.py
│   ├── indexing.py
│   ├── reshape.py
│   ├── series.py
│   └── window.py
├── io/
├── tests/
└── utils/
```

Essentially:

* The `core/` directory contains the implementation of the core classes `Index`, `Series` and `DataFrame` and its methods. Other "non-core" directories include `io`, `tests`, `utils` and in the future perhaps `plotting` and `gis` (?).

* A module (e.g., `window.py`) can be promoted to a package (e.g., `window/`) if its complexity grows to the point that it is better organized into multiple files

**Comment by shwina** [comment]
Closed by #2648
